src/porepy/numerics/ad/operators.py

```diff
@@ -872,12 +882,162 @@ def parse(self, mdg: pp.MixedDimensionalGrid) -> np.ndarray:
+    in data[pp.STATE][pp.ITERATE][self._name] for the current time and
+    data[pp.STATE][self._name] for the previous time.
+
+    The array can be differentiated in time using pp.ad.dt().
+
+    The intended use is to represent time-varying quantities in equations, e.g., source
+    terms. Future use will also include numerical values of boundary conditions,
+    however, this is pending an update to the model classes.
+
+    Attributes:
+        prev_time (boolean): If True, the the array will be evaluated using in
+            data[pp.STATE] (data being the data dictionaries for subdomains and
+            interfaces), if False, data[pp.STATE][pp.ITERATE] is used.
+
+    """
+
+    def __init__(
+        self,
+        name: str,
+        subdomains: Optional[list[pp.Grid]] = None,
+        interfaces: Optional[list[pp.MortarGrid]] = None,
+        previous_timestep: bool = False,
+    ):
+        """Initialize a TimeDependentArray.
+
+        Args:
+            name: Name of the variable. Should correspond to items in data[pp.STATE].
+            subdomains: Subdomains on which the array is defined. Defaults to None.
+            interfaces: Interfaces on which the array is defined. Defaults to None.
+
+            Exactly one of subdomains and interfaces must be non-empty.
+
+            previous_timestep: Flag indicating if the array should be evaluated at the
+                previous time step.
+
+        Raises:
+            ValueError: If either none of, or both of, subdomains and interfaces are
+                empty.
+
+        """
+
+        self._name: str = name
+
+        if subdomains is None:
+            subdomains = []
+        if interfaces is None:
+            interfaces = []
+
+        if len(interfaces) == 0 and len(subdomains) == 0:
+            raise ValueError(
+                "A time dependent array must be associated with an"
+                " interface or a subdomain."
+            )
+        if len(interfaces) > 0 and len(subdomains) > 0:
+            raise ValueError(
+                "A time dependent array must be associated with either an"
+                " interface or a subdomain."
+            )
+
+        # Store subdomains and interfaces.
+        self.subdomains = subdomains
+        self.interfaces = interfaces
+        self._g: Sequence[GridLike]
+        self._is_interface_arary: bool
+
+        # Shorthand access to grid or edge:
+        if len(interfaces) == 0:
+            # Appease mypy
+            assert all([isinstance(sd, pp.Grid) for sd in subdomains])
+            self._g = subdomains
+            self._is_interface_arary = False
+        else:
+            assert all([isinstance(intf, pp.MortarGrid) for intf in interfaces])
+            self._g = interfaces
+            self._is_interface_arary = True
+
+        self.prev_time: bool = previous_timestep
+
+        self._set_tree()
+
+    def previous_timestep(self) -> TimeDependentArray:
```

Comment:
I would prefer not to use it in `Operator,` since that would make it unclear which variables can actually be taken at the previous time step ( I can see this leading to a lot of confusion, some of it will arguably be the user's fault, but still).

A mixin won't work either, since the return type of `previous_timestep()` (and the signature of the various classes) varies. This also implies that adding time dependency to the Operator class does not help much - this method must still be implemented.

Reply:
There are workarounds for different signatures and return types, but those are typically rather advanced (and opaque)

Reply:
I know. All things considered, I suggest we keep this for now. We may revisit, though, for instance if we get a better system for dealing with different states and iterates, e.g. in connection with an expansion of the solution algorithms.